### PR TITLE
lightning: support data KV checking of 'replace' mode for lightning post-import conflict detection

### LINKED ISSUE #45774: Support `replace` mode for lightning post-import conflict detection
## Feature Request

**Is your feature request related to a problem? Please describe:**

Currently lightning only supports `remove` mode for post-import conflict detection, but many customers request lightning to support `replace` mode for lightning post-import conflict detection.

**Describe the feature you'd like:**

`replace` mode for lightning post-import conflict detection:
To resolve rows with conflict, instead of deleting all the rows that are engaged in conflict (the algorithm for `remove`), we delete some rows with conflict and reserve other rows that can be kept and not cause conflict anymore. Under this circumstance, we only delete the necessary rows to resolve conflicts, so that we can keep more original rows than `remove` mode as long as the conflicts are resolved. 

For instance, we have the following table structure:

```
create table test (c1 int primary key, c2 int);
```

And we have 2 rows:

```
1, 111
1, 222
```

`remove` mode would delete both 2 rows, but `replace` mode would only delete 1 row and reserve the other one.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

Design documentation:

https://pingcap.feishu.cn/docx/XD52dwxfdoewUQxDDRtcdUq4nod?from=from_copylink

**Implementation steps:**

1. Change the structure of the tables that record data KV conflicts and index KV conflicts. Add index to track and search rows with conflict faster.
https://github.com/pingcap/tidb/blob/1e73eee4860cc3eb91c031bc943ea6a75bc1a717/br/pkg/lightning/backend/local/duplicate.go#L710-L714

2. Use local backend import, set `duplicate-resolution = 'record'` to record the conflict rows but not delete them from downstream, manually use TiKV's API to extract information from the tables, and implement the `replace` algorithm to check the feasibility of the method.

3. Modify `ResolveAllConflictKeys` function to support the `replace` algorithm that delete only the necessary rows to resolve conflict.
https://github.com/pingcap/tidb/blob/1e73eee4860cc3eb91c031bc943ea6a75bc1a717/br/pkg/lightning/backend/local/duplicate.go#L1012

4. Optimize executing logic and performance for  the `replace` algorithm.

### PR DESCRIPTION
**Pull request by lyzx2001**
### What problem does this PR solve?

Issue Number: close #45774 

Problem Summary:

Currently lightning only supports "remove" mode for post-import conflict detection, but many customers request lightning to support "replace" mode for lightning post-import conflict detection.

We would like to support "replace" mode for lightning post-import conflict detection:
To resolve rows with conflict, instead of deleting all the rows that are engaged in conflict (the algorithm for remove), we delete some of the rows with conflict and reserve other rows that can be kept and not cause conflict anymore. Under this circumstance, we only delete the necessary rows to resolve conflicts, so that we can keep more original rows than remove mode as long as the conflicts are resolved.

The algorithms for index KV checking is contained in https://github.com/pingcap/tidb/pull/45926

This PR contains the algorithms for data KV checking.

### What is changed and how it works?

Demo code for 'replace' mode of lightning post-import conflict detection:

https://github.com/lyzx2001/tidb-conflict-replace

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Lightning post-import conflict detection has supported "replace" mode, which is to delete some rows with conflict and reserve other rows that can be kept and not cause conflict anymore.

In comparison, the algorithm for "remove" mode is to delete all the rows that are engaged in conflict.
```

### CODE REVIEW

File: br/pkg/lightning/errormanager/errormanager.go [hunk 1]
```diff
@@ -148,6 +148,20 @@ const (
 		ORDER BY raw_key;
 	`
 
+	selectDataConflictKeysReplace = `
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:151]: same problem as https://github.com/pingcap/tidb/pull/45926#discussion_r1302463762

File: br/pkg/lightning/errormanager/errormanager.go [hunk 2]
```diff
@@ -642,11 +656,134 @@ func (em *ErrorManager) ReplaceConflictKeys(
...
+				}
+
+				// find out all the KV pairs that are contained in the data KV
+				kvPairs := encoder.SessionCtx.TakeKvPairs()
+				for _, kvPair := range kvPairs.Pairs {
+					em.logger.Debug("got encoded KV",
+						logutil.Key("key", kvPair.Key),
+						zap.Binary("value", kvPair.Val))
+					kvLatestValue, err := fnGetLatest(gCtx, kvPair.Key)
+					if tikverr.IsErrNotFound(err) || kvLatestValue == nil {
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:746]: should we check `IsErrNotFound` for other usage of `fnGetLatest`?

File: br/pkg/lightning/errormanager/errormanager.go [hunk 3]
```diff
@@ -642,11 +656,134 @@ func (em *ErrorManager) ReplaceConflictKeys(
...
+
+					// if the value of the KV pair is not equal to the latest value of the key of the KV pair
+					// that means the value of the KV pair has been overwritten, so it needs no extra operation
+					if !bytes.Equal(kvLatestValue, kvPair.Val) {
+						continue
+					}
+
+					// if the KV pair is contained in mustKeepKvPairs, we cannot delete it
+					// if not, delete the KV pair
+					isContained := false
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:761]: try slices.ContainsFunc

File: br/pkg/lightning/errormanager/errormanager_test.go [hunk 4]
```diff
@@ -248,6 +248,42 @@ func TestReplaceConflictKeysIndexKvChecking(t *testing.T) {
...
+		types.NewIntDatum(6),
+		types.NewStringDatum("2.csv"),
+	}
+	_, err = encoder.Table.AddRecord(encoder.SessionCtx, data1)
+	require.NoError(t, err)
+	_, err = encoder.Table.AddRecord(encoder.SessionCtx, data2)
+	require.NoError(t, err)
+	kvPairs := encoder.SessionCtx.TakeKvPairs()
+
+	rawKey := kvPairs.Pairs[1].Key
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager_test.go:280]: I'd like to rename these variables like `data1RowKey`, `data2IndexValue` so reviewers can easier to remember their meaning.

File: br/pkg/lightning/errormanager/errormanager_test.go [hunk 5]
```diff
@@ -322,6 +342,204 @@ func TestReplaceConflictKeysIndexKvChecking(t *testing.T) {
...
+	require.NoError(t, err)
+	_, err = encoder.Table.AddRecord(encoder.SessionCtx, data3)
+	require.NoError(t, err)
+	_, err = encoder.Table.AddRecord(encoder.SessionCtx, data4)
+	require.NoError(t, err)
+	_, err = encoder.Table.AddRecord(encoder.SessionCtx, data5)
+	require.NoError(t, err)
+	kvPairs := encoder.SessionCtx.TakeKvPairs()
+
+	rawKey1 := kvPairs.Pairs[7].Key
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager_test.go:457]: rename like above comment

File: br/pkg/lightning/errormanager/errormanager.go [hunk 6]
```diff
@@ -148,6 +148,20 @@ const (
 		ORDER BY raw_key;
 	`
 
+	selectDataConflictKeysReplace = `
+		SELECT raw_key
+		FROM %s.` + ConflictErrorTableName + `
+		WHERE table_name = ? AND index_name = 'PRIMARY'
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:154]: and seems we should add more INDEX covering the WHERE clause to `ConflictErrorTableName`

**Review by lance6716 — COMMENTED** [review]
I'd like to add more test in this PR or in next PR.
For unit tests, the criteria is to improve the coverage. It's helpful to force us to think which corner cases can cover the code path.
For integration tests, we will learn if related API behaviour is expected, like what's the error of "key not exist" using fnGetLatest

File: br/pkg/lightning/errormanager/errormanager.go (hunk 1 above)
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:152]: I think here the situation is different than https://github.com/pingcap/tidb/pull/45926#discussion_r1302463762

We need to first use `rawKey` to determine `mustKeepKvPairs` for this `rawKey`, then iterate all the KV with this `rawKey`, in the iteration this corresponding `mustKeepKvPairs` is needed. If we just use one SQL query, we may need to store all `mustKeepKvPairs`, which is not efficient.

File: br/pkg/lightning/errormanager/errormanager.go (hunk 2 above)
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:747]: I don't think so. Other usage of `fnGetLatest` will not get the result that the key does not exist, if so, we directly return  error.

File: br/pkg/lightning/errormanager/errormanager.go [hunk 7]
```diff
@@ -94,7 +95,8 @@ const (
 			raw_handle  mediumblob NOT NULL COMMENT 'the data handle derived from the conflicted key or value',
 			raw_row     mediumblob NOT NULL COMMENT 'the data retrieved from the handle',
 			KEY (task_id, table_name),
-			INDEX (index_name)
+			INDEX (index_name),
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:98]: just noticed in line 97 we use KEY as the keyword for index, for one style we should unify them

File: br/pkg/lightning/errormanager/errormanager.go [hunk 8]
```diff
@@ -642,11 +651,116 @@ func (em *ErrorManager) ReplaceConflictKeys(
...
+			}
+
+			// find out all the KV pairs that are contained in the data KV
+			kvPairs := encoder.SessionCtx.TakeKvPairs()
+			for _, kvPair := range kvPairs.Pairs {
+				em.logger.Debug("got encoded KV",
+					logutil.Key("key", kvPair.Key),
+					zap.Binary("value", kvPair.Val))
+				kvLatestValue, err := fnGetLatest(gCtx, kvPair.Key)
+				if tikverr.IsErrNotFound(err) || kvLatestValue == nil {
```
Comment (lance6716) [inline br/pkg/lightning/errormanager/errormanager.go:734]: seems `fnGetLatest` will return error which will be checked by `IsErrNotFound`. For above usage of `fnGetLatest` do we need to ignore the not-found error and continue the function?

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: br/pkg/lightning/errormanager/errormanager.go (hunk 8 above)
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:734]: according to the demo code, here if the value of `kvPair.Key` is not found, that means `kvPair.Key` has been deleted, so we need to continue in the for loop
https://github.com/lyzx2001/tidb-conflict-replace/blob/9666d86f660f9a91e4a0446027bf0319a88abf2a/src/main.go#L178-L180

Comment (lance6716) [reply br/pkg/lightning/errormanager/errormanager.go:734]: Why do we return for not-found error in line 863

Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:734]: thanks for reminding, we should first check whether `err` is not-found error

File: br/tests/lightning_duplicate_resolution_replace/data/dup_resolve.a.1.csv [hunk 9]
```diff
@@ -2,3 +2,5 @@ a,b,c
 1,6,1.csv
 2,6,2.csv
 3,3,3.csv
+3,4,4.csv
+5,4,5.csv
```
Comment (lance6716) [inline br/tests/lightning_duplicate_resolution_replace/data/dup_resolve.a.1.csv:6]: the new rows should also be checked in run.sh

**Review by lance6716 — COMMENTED** [review]
rest lgtm. We can add more complex test case and improve the code structure / performance in following PR

File: br/pkg/lightning/errormanager/errormanager_test.go [hunk 10]
```diff
@@ -261,28 +297,201 @@ func TestReplaceConflictKeysIndexKvChecking(t *testing.T) {
...
+				Name:   model.NewCIStr("b"),
+				Offset: 1,
+				Length: -1,
+			}},
+		Unique:  true,
+		Primary: false,
+		State:   model.StatePublic,
+	}
+
+	table := &model.TableInfo{
```
Comment (D3Hunter) [inline br/pkg/lightning/errormanager/errormanager_test.go:393]: can use this to mock table info from sql

https://github.com/pingcap/tidb/blob/2680a9a8ea148c5dfefe046512b587bc8880816d/disttask/importinto/encode_and_sort_operator_test.go#L190C1-L194C71

**Review by D3Hunter — COMMENTED** [review]
rest lgtm

File: br/tests/lightning_duplicate_resolution_replace/data/dup_resolve.a.1.csv (hunk 9 above)
Comment (lyzx2001) [reply br/tests/lightning_duplicate_resolution_replace/data/dup_resolve.a.1.csv:6]: row 4 was deleted because of conflict with row 3, and row 5 was deleted because of conflict with row 4, so they should not appear in `dup_resolve.a`

File: br/pkg/lightning/errormanager/errormanager_test.go (hunk 10 above)
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager_test.go:393]: we need to use `ddl.MockTableInfo` for this method, but if we import `github.com/pingcap/tidb/ddl`, it would result in "import cycle not allowed in test"

```
# github.com/pingcap/tidb/br/pkg/lightning/errormanager
FAIL	github.com/pingcap/tidb/br/pkg/lightning/errormanager [setup failed]
package github.com/pingcap/tidb/br/pkg/lightning/errormanager
	imports github.com/pingcap/tidb/ddl
	imports github.com/pingcap/tidb/ddl/ingest
	imports github.com/pingcap/tidb/br/pkg/lightning/errormanager: import cycle not allowed in test
FAIL
````

File: br/pkg/lightning/errormanager/errormanager.go [hunk 11]
```diff
@@ -642,11 +651,116 @@ func (em *ErrorManager) ReplaceConflictKeys(
...
+					// calculate the new mustKeepKvPairs corresponding to the new rawKey
+					// find out all the KV pairs that are contained in the data KV
+					mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
+				}
+			}
 
-		// TODO: check data KV
+			// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
+			// if not, that means this data KV has been deleted due to overwritten index KV
+			if bytes.Equal(rawValue, latestValue) {
```
[the remark this answers is not in the document]
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:709]: we only need to decode `latestValue` when `!bytes.Equal(rawKey, previousRawKey)`, so if we put this check before we decode `latestValue`, we need to check `!bytes.Equal(rawKey, previousRawKey)` again to determine whether we need to decode `latestValue`, which seems unnecessary

[the remark this answers is not in the document]
Comment (D3Hunter) [reply br/pkg/lightning/errormanager/errormanager.go:709]: no need to check twice, like this:
```go
			if bytes.Equal(rawKey, previousRawKey) {
				continue
			}
			previousRawKey = rawKey
			// get the latest value of rawKey from downstream TiDB
			latestValue, err = fnGetLatest(gCtx, rawKey)
			if err != nil && !tikverr.IsErrNotFound(err) {
				return errors.Trace(err)
			}

			// TODO: check data KV				// if the latest value of rawKey equals to rawValue, that means this data KV is maintained in downstream TiDB
			// if not, that means this data KV has been deleted due to overwritten index KV
			if bytes.Equal(rawValue, latestValue) {
				continue
			}
			if latestValue != nil {
				handle, err := tablecodec.DecodeRowKey(rawKey)
				if err != nil {
					return errors.Trace(err)
				}
				decodedData, _, err := tables.DecodeRawRowData(encoder.SessionCtx,
					tbl.Meta(), handle, tbl.Cols(), latestValue)
				if err != nil {
					return errors.Trace(err)
				}
				_, err = encoder.Table.AddRecord(encoder.SessionCtx, decodedData)
				if err != nil {
					return errors.Trace(err)
				}
				// calculate the new mustKeepKvPairs corresponding to the new rawKey
				// find out all the KV pairs that are contained in the data KV
				mustKeepKvPairs = encoder.SessionCtx.TakeKvPairs()
			}
```

[the remark this answers is not in the document]
Comment (lyzx2001) [reply br/pkg/lightning/errormanager/errormanager.go:709]: we cannot directly write this 
```go
if bytes.Equal(rawKey, previousRawKey) {
	continue
}
```
because if rawKey equals to previousRawKey, we still need to conduct the code from line 707 to line 763

**Review by D3Hunter — APPROVED** [review]

**Review by lance6716 — APPROVED** [review]

**Review by 3pointer — APPROVED** [review]
